fix header parser single quote bug

demerphq opened the pull request:
HeaderParser.pm - parse single quoted backslash properly
    
@khwilliamson  gave me a test script which was breaking because we did not handle single quoted escapes like '\\' and '\n' and the  like. This fixes the tokenizer pattern to handle these without dieing, although we do not validate that the escape sequence itself is legal, we no longer choke on having backslash something in the single quote.

This patch sequence also includes a commit which teaches t/porting/header_parser.t how to update its own Data::Dumper based test. Updating this test manually was annoying, so I made the script do it for us when needed. 

* This set of changes does not require a perldelta entry.

mauke commented on regen/HeaderParser.pm:
```diff
@@ -115,7 +115,7 @@ BEGIN {
                 (?<define> defined\(\w+\) )
             |   (?<func>   \w+\s*\(\s*\w+(?:\s*,\s*\w+)*\s*\) )
-            |   (?<const>  (?:0x[a-fA-F0-9]+|\d+[LU]*|'.') )
+            |   (?<const>  (?:0x[a-fA-F0-9]+|\d+[LU]*|'.'|'\\.') )
```
I'm not a fan of `'.'`. It matches `'''` and `'\'`, neither of which is valid C.

Also, multi-char constants are allowed: `'AB'` is valid C (but IIRC the value of this constant is implementation-defined.)

(Also also, the `[LU]` numeric suffix should be `[LUlu]` because it's case insensitive.)

We should do something like for string literals:
```
(?<const> (?: ... |'(?:[^'\\]|\\.)+') )
```

demerphq replied:
I don't mind looking into these improvements as a separate patch, but I dont know when ill get time for it and im a little concerned we are opening a can of worms. The purpose of HeaderParser isnt to completely parse C, but rather to parse enough C pre-processor that we can do textual manipulation of it without horribly breaking things. So for now i'd rather we applied this part as is to get Karl working and then follow up with more validation of constructs like you mention.

khwilliamson replied:
Agreed.  This is not a general purpose parser that we release for public use.  It should be able to parse all the core hdr files including constructs that are likely to be added.  I'm actually wanting this for use in Devel::PPPort, which parses header files for the current and previous Perl versions during its regeneration for each new release of the module.  The current parser doesn't do nearly as good a job as this one.  So multi-char constants may be valid C, but they aren't something we are going to use.  Uu and string constants might be.

mauke replied:
I mean, if you don't want to support multi-char constants, just remove the `+` from `'(?:[^'\\]|\\.)+'`. And if you don't want to factor out the common `'` matches, you can write it in two parts: `'[^'\\]'|'\\.'`
But widening `'[^'\\]'` to `'.'` (as in the original code) is too sloppy to feel comfortable for me.

demerphq replied:
I did not say we should not do what you suggest, I said I thought it could be a can of worms, and that we should address it in a follow up patch. This MR is meant to be a quick fix to get Karl up and running with something he is working on.  We can then thoughtfully review and improve this logic along the lines you suggest in the coming days and weeks.

demerphq replied:
Anyway, it turned out I had a bit of time today to work on this. So its fixed now.

mauke commented on t/porting/header_parser.t:
```diff
@@ -55,10 +55,37 @@ $hp->parse_text(<<~'EOF');
 my $normal= $hp->lines_as_str();
 my $lines= $hp->lines();
 my $lines_as_str= Data::Dumper->new([$lines])->Sortkeys(1)->Useqq(1)->Indent(1)->Dump();
+if ($ENV{UPDATE_DUMP}) {
```
`UPDATE_DUMP` is a bit unspecific as a name. Do you think context is enough to deal with this (and no perl developer will ever have `UPDATE_DUMP=1` in their environment for other reasons), or should we switch to something like `PERL_DEV_UPDATE_DUMP`?

demerphq replied:
I doubt UPDATE_DUMP will be in place when someone runs this test file, but I dont mind changing the var either. Ill do that now.

demerphq committed:
regen/HeaderParser.pm - fix various edge cases

We were not handling things like '\\' '\x01' nor 'ABCD' (multichar
constants), nor were we handling spaces in expressions like "defined
(foo)", also we weren't testing that we were warning free.

This patch also includes support for auto-updating the two larger tests
in t/porting/header_parser.t, now if you set the env var
PERL_DEV_UPDATE_TESTFILE to a true value and run the script it will
autoupdate the script with the latest output that is generated.

demerphq committed:
HeaderParser.pm - auto indent defines in conditional clauses

There are constructor options to control this, but we default
to on.
